sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/util/ReduceFnRunner.java

```diff
@@ -710,6 +717,18 @@ private void emitIfAppropriate(ReduceFn<K, InputT, OutputT, W>.Context directCon
       paneInfoTracker.clear(directContext.state());
       activeWindows.remove(directContext.window());
     }
+
+    if (shouldDiscard &&
+        !triggerRunner.hasState(directContext.window(), directContext.timers(),
```

Comment:
Jumping in here without looking deeper, I think the trigger's state should always be a summary of the pending pane, so should be able to end up cleared perhaps more than this. Only finished bits and orFinally might change this. Apologies if my high level scan missed the details. Will look deeper later.